pypdf/_page.py

```diff
@@ -460,12 +460,94 @@ def images(self) -> List[File]:
         for obj in x_object:
             if x_object[obj][IA.SUBTYPE] == "/Image":
-                extension, byte_stream = _xobj_to_image(x_object[obj])
+                extension, byte_stream, img = _xobj_to_image(x_object[obj])
                 if extension is not None:
                     filename = f"{obj[1:]}{extension}"
                     images_extracted.append(File(name=filename, data=byte_stream))
+                    images_extracted[-1].image = img
+                    images_extracted[-1].indirect_reference = x_object[
+                        obj
+                    ].indirect_reference
         return images_extracted
 
+    def _get_ids_image(
+        self, obj: Optional[DictionaryObject] = None, ancest: Optional[List[str]] = None
+    ) -> List[Union[str, List[str]]]:
+        if obj is None:
+            obj = self
+        if ancest is None:
+            ancest = []
+        lst: List[Union[str, List[str]]] = []
+        if RES.XOBJECT not in obj[PG.RESOURCES]:  # type: ignore
+            return lst
+
+        x_object = obj[PG.RESOURCES][RES.XOBJECT].get_object()  # type: ignore
+        for o in x_object:
+            if x_object[o][IA.SUBTYPE] == "/Image":
+                lst.append(o if len(ancest) == 0 else ancest + [o])
+            else:  # is a form with possible images inside
+                lst.extend(self._get_ids_image(x_object[o], ancest + [o]))
+        return lst  # type: ignore
+
+    def _get_image(
+        self,
+        id: Union[str, List[str], Tuple[str]],
+        obj: Optional[DictionaryObject] = None,
+    ) -> File:
+        if obj is None:
+            obj = cast(DictionaryObject, self)
+        if isinstance(id, tuple):
+            id = list(id)
+        if isinstance(id, List) and len(id) == 1:
+            id = id[0]
+        xobjs = cast(
+            DictionaryObject, cast(DictionaryObject, obj[PG.RESOURCES])[RES.XOBJECT]
+        )
+        if isinstance(id, str):
+            imgd = _xobj_to_image(cast(DictionaryObject, xobjs[id]))
+            extension, byte_stream = imgd[:2]
+            f = File(
+                name=f"{id[1:]}{extension}",
+                data=byte_stream,
+                image=imgd[2],
+                indirect_reference=xobjs[id].indirect_reference,
+            )
+            return f
+        else:  # in a sub object
+            ids = id[1:]
+            return self._get_image(ids, cast(DictionaryObject, xobjs[id[0]]))
+
+    @property
+    def images(self) -> List[File]:
+        """
+            Read-only property that emulates a list of files
+            Get a list of all images of the page.
+
+            the key can be:
+              µan str (for top object) or a tuple for image within XObject forms
+              or an int
+        ex:
+        ```
+        reader.pages[0].images[0]        # return fist image
+        reader.pages[0].images['/I0']    # return image '/I0'
```

Comment:
Can it happen that there are name clashes, e.g. that there are two different images with the same key?

Reply:
no : it is a dictionary actually.
The only case you can get is that the name is identical within two XObject/Forms with different names (the tuple / sequence will be different)